formats/opentype.fathom

```diff
@@ -1036,6 +1038,527 @@ let extension_substitution = (
+    let X_ADVANCE : U16 = 0x0004;
+    let Y_ADVANCE : U16 = 0x0008;
+    let X_PLACEMENT_DEVICE : U16 = 0x0010;
+    let Y_PLACEMENT_DEVICE : U16 = 0x0020;
+    let X_ADVANCE_DEVICE : U16 = 0x0040;
+    let Y_ADVANCE_DEVICE : U16 = 0x0080;
+
+    let optional_field = fun (field : U16) => fun (format : Format) =>
+        match (u16_neq (u16_and flags field) 0) {
+            true => format,
+            false => empty,
+        };
+
+    {
+        /// Horizontal adjustment for placement, in design units.
+        x_placement <- optional_field X_PLACEMENT s16be,
+        /// Vertical adjustment for placement, in design units.
+        y_placement <- optional_field Y_PLACEMENT s16be,
+        /// Horizontal adjustment for advance, in design units — only used for horizontal layout.
+        x_advance <- optional_field X_ADVANCE s16be,
+        /// Vertical adjustment for advance, in design units — only used for vertical layout.
+        y_advance <- optional_field Y_ADVANCE s16be,
+        /// Offset to Device table (non-variable font) / VariationIndex table (variable font) for
+        /// horizontal placement, from beginning of the immediate parent table (SinglePos or
+        /// PairPosFormat2 lookup subtable, PairSet table within a PairPosFormat1 lookup subtable)
+        /// — may be NULL.
+        x_pla_device_offset <- optional_field X_PLACEMENT_DEVICE (offset16 table_start device_or_variation_index_table),
+        /// Offset to Device table (non-variable font) / VariationIndex table (variable font) for
+        /// vertical placement, from beginning of the immediate parent table (SinglePos or
+        /// PairPosFormat2 lookup subtable, PairSet table within a PairPosFormat1 lookup subtable)
+        /// — may be NULL.
+        y_pla_device_offset <- optional_field Y_PLACEMENT_DEVICE (offset16 table_start device_or_variation_index_table),
+        /// Offset to Device table (non-variable font) / VariationIndex table (variable font) for
+        /// horizontal advance, from beginning of the immediate parent table (SinglePos or
+        /// PairPosFormat2 lookup subtable, PairSet table within a PairPosFormat1 lookup subtable)
+        /// — may be NULL.
+        x_adv_device_offset <- optional_field X_ADVANCE_DEVICE (offset16 table_start device_or_variation_index_table),
+        /// Offset to Device table (non-variable font) / VariationIndex table (variable font) for
+        /// vertical advance, from beginning of the immediate parent table (SinglePos or
+        /// PairPosFormat2 lookup subtable, PairSet table within a PairPosFormat1 lookup subtable)
+        /// — may be NULL.
+        y_adv_device_offset <- optional_field Y_ADVANCE_DEVICE (offset16 table_start device_or_variation_index_table),
+    }
+);
+
+/// A value record that is `empty` if flags is 0
+let optional_value_record = fun (table_start : Pos) => fun (flags : U16) => match (u16_eq flags 0) {
+    true => empty,
+    false => value_record table_start flags
+};
```

Comment:
Won't `value_record` already be empty if `flags` is zero?

Reply:
Yep. I initially was going to just use `value_record` but doing that results in something like the following in the output:

```
{
    x_placement <- empty,
    y_placement <- empty,
    x_advance <- empty,
    y_advance <- empty,
    x_pla_device_offset <- empty,
    y_pla_device_offset <- empty,
    x_adv_device_offset <- empty,
    y_adv_device_offset <- empty,
}
```

instead of just `empty`. I thought having it named `optional_value_record` at the call-site was useful for clarity too.